src/core/Events.js

```diff
@@ -216,22 +203,61 @@ export var Events = {
 	},
 
 	// @method listens(type: String, propagate?: Boolean): Boolean
+	// @method listens(type: String, fn: Function, context?: Object, propagate?: Boolean): Boolean
 	// Returns `true` if a particular event type has any listeners attached to it.
 	// The verification can optionally be propagated, it will return `true` if parents have the listener attached to it.
-	listens: function (type, propagate) {
+	listens: function (type, fn, context, propagate) {
 		if (typeof type !== 'string') {
 			console.warn('"string" type argument expected');
 		}
+
+		if (typeof fn !== 'function') {
+			propagate = !!fn;
+			fn = undefined;
+			context = undefined;
+		}
+
 		var listeners = this._events && this._events[type];
-		if (listeners && listeners.length) { return true; }
+		if (listeners && listeners.length) {
+			if (this._listens(type, fn, context)) {
+				return true;
+			}
+		}
 
 		if (propagate) {
 			// also check parents for listeners if event propagates
 			for (var id in this._eventParents) {
-				if (this._eventParents[id].listens(type, propagate)) { return true; }
+				if (this._eventParents[id].listens(type, fn, context, propagate)) { return true; }
+			}
+		}
+		return false;
+	},
+
+	_listens: function (type, fn, context) {
+		if (!this._events) {
+			return false;
+		}
+
+		var listeners = this._events[type] || [];
+		if (!fn) {
+			return !!listeners.length;
+		}
+
+		if (context === this) {
+			// Less memory footprint.
+			context = undefined;
+		}
+
+		for (var i = 0, len = listeners.length; i < len; i++) {
+			if (listeners[i].fn === fn && listeners[i].ctx === context) {
+				return {
+					index: i,
+					listener: listeners[i]
+				};

```

Comment:
The problem with this approach is that while it's certainly more readable, it has an overhead of an extra object allocation for the return result. It might not be significant, but it's still overhead and if we can avoid it, I think we should.